src/Utils/MathCudaKernels.cu

```diff
@@ -0,0 +1,115 @@
+// EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Additional BSD Notice
+// 1. This notice is required to be provided under our contract with the U.S. Department
+// of Energy (DOE). This work was produced at Lawrence Livermore National Laboratory under
+// Contract No. DE-AC52-07NA27344 with the DOE.
+// 2. Neither the United States Government nor Lawrence Livermore National Security, LLC
+// nor any of their employees, makes any warranty, express or implied, or assumes any
+// liability or responsibility for the accuracy, completeness, or usefulness of any
+// information, apparatus, product, or process disclosed, or represents that its use would
+// not infringe privately-owned rights.
+// 3. Also, reference herein to any specific commercial products, process, or services by
+// trade name, trademark, manufacturer or otherwise does not necessarily constitute or
+// imply its endorsement, recommendation, or favoring by the United States Government or
+// Lawrence Livermore National Security, LLC. The views and opinions of authors expressed
+// herein do not necessarily state or reflect those of the United States Government or
+// Lawrence Livermore National Security, LLC, and shall not be used for advertising or
+// product endorsement purposes.
+
+/**
+ * @file MathCudaKernels.cu
+ *
+ * @author Cosmin G. Petra <petra1@llnl.gov>, LNNL
+ * @author Nai-Yuan Chiang <chiang7@llnl.gov>, LNNL
+ *
+ */
+
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <cuda.h>
+#include <curand_kernel.h>
+#include <curand.h>
+
+#include <cuda_runtime.h>
+#include <device_launch_parameters.h>
+#include "hiopCppStdUtils.hpp"
+#include "MathDeviceKernels.hpp"
+
+__global__
+void array_random_uniform_cuda(int n, double* d_array, unsigned long seed, double minv, double maxv)
+{
+    const int num_threads = blockDim.x * gridDim.x;
+    const int tid = blockIdx.x * blockDim.x + threadIdx.x;    
+    double ranv;
+    curandState state;
+    curand_init( seed, tid, 0, &state);
+    for (int i = tid; i < n; i += num_threads) {
+      ranv = curand_uniform_double( &state ); // from 0 to 1
+      d_array[i] = ranv * (maxv - minv) + minv;	
+    }
+}
+
+namespace hiop
+{
+namespace device
+{
+
+int array_random_uniform_kernel(int n, double* d_array, double minv, double maxv)
+{
+  int block_size=256;
+  int grid_size = (n+block_size-1)/block_size;
+  
+  unsigned long seed = generate_seed();
+  array_random_uniform_cuda<<<grid_size,block_size>>>(n, d_array, seed, minv, maxv);
+
+  return 1;
+}
+
+int array_random_uniform_kernel(int n, double* d_array)
+{
+  unsigned long seed = generate_seed();
+
+  curandGenerator_t generator;
+  curandCreateGenerator(&generator, CURAND_RNG_PSEUDO_DEFAULT);
+  curandSetPseudoRandomGeneratorSeed(generator, seed);
+
+  // generate random val from 0 to 1
+  curandGenerateUniformDouble(generator, d_array, n);
+
+  curandDestroyGenerator(generator);
+  return 1;
+}
```

Comment:
Not sure which implementation is faster. The second method will adjust the values in the outer layer [(see here)](https://github.com/LLNL/hiop/blob/be9c984ed928b3a29f809f85f37fc9fabcfbf1b5/src/LinAlg/hiopVectorRajaPar.cpp#L230-L236)
@cnpetra @CameronRutherford any experience?

Reply:
use the one with 4 arguments -- it "fuses" kernels